core/src/main/java/org/apache/calcite/util/Util.java

```diff
@@ -1717,16 +1719,14 @@ private static void appendPosixTime(StringBuilder buf, int millis) {
    * @return Java locale object
    */
   public static Locale parseLocale(String localeString) {
-    String[] strings = localeString.split("_");
-    switch (strings.length) {
-    case 1:
-      return new Locale(strings[0]);
-    case 2:
-      return new Locale(strings[0], strings[1]);
-    case 3:
-      return new Locale(strings[0], strings[1], strings[2]);
-    default:
-      throw new AssertionError("bad locale string '" + localeString + "'");
+    if (localeString.isEmpty()) {
+      return Locale.ROOT;
+    }
+    try {
+      return new Locale.Builder().setLanguageTag(
+          UNDERSCORE.matcher(localeString).replaceAll("-")).build();
+    } catch (IllformedLocaleException e) {
+      throw new AssertionError("bad locale string '" + localeString + "'", e);
```

Comment:
A `IllegalStateException` or `IllegalArgumentException` would be more appropriate here but since it was an `AssertionError` before I am also fine leaving it as is.
I don't think an `AssertionError` is appropriate cause we may arrive here by user input.

Reply:
makes sense, thanks